libs/remix-ui/run-tab/src/lib/components/instanceContainerUI.tsx

```diff
@@ -17,14 +17,37 @@ export function InstanceContainerUI(props: InstanceContainerProps) {
       if (enableSave.current) {
         const { network } = await props.plugin.call('blockchain', 'getCurrentNetworkStatus')
         chainId.current = network.id
+        // Move contract saved in localstorage to Remix FE
         const allSavedContracts = localStorage.getItem('savedContracts')
         if (allSavedContracts) {
-          await props.plugin.call('udapp', 'clearAllSavedInstances')
           const savedContracts = JSON.parse(allSavedContracts)
-          if (savedContracts && savedContracts[network.id]) {
-            const instances = savedContracts[network.id]
-            for (const inst of instances)
-              if (inst) await props.plugin.call('udapp', 'addSavedInstance', inst.address, inst.abi || inst.contractData.abi, inst.name, inst.savedOn, inst.filePath)
+          for (const networkId in savedContracts) {
+            if (savedContracts[networkId].length > 0) {
+              for (const contractDetails of savedContracts[networkId]) {
+                const objToSave = {
+                  name: contractDetails.name,
+                  address: contractDetails.address,
+                  abi: contractDetails.abi || contractDetails.contractData.abi,
+                  filePath: contractDetails.filePath,
+                  pinnedAt: contractDetails.savedOn
+                }
+                await props.plugin.call('fileManager', 'writeFile', `.deploys/pinned-contracts/${networkId}/${contractDetails.address}.json`, JSON.stringify(objToSave, null, 2))
+              }
+            }
+          }
+          localStorage.removeItem('savedContracts')
+        }
+        // Clear existing saved instance state
+        await props.plugin.call('udapp', 'clearAllSavedInstances')
+        // Load contracts from FE
+        const isPinnedAvailable = await props.plugin.call('fileManager', 'exists', `.deploys/pinned-contracts/${chainId.current}`)
+        if (isPinnedAvailable) {
+          const list = await props.plugin.call('fileManager', 'readdir', `.deploys/pinned-contracts/${chainId.current}`)
+          const filePaths = Object.keys(list)
+          for (const file of filePaths) {
+            const pinnedContract = await props.plugin.call('fileManager', 'readFile', file)
+            const pinnedContractObj = JSON.parse(pinnedContract)
+            if (pinnedContractObj) await props.plugin.call('udapp', 'addSavedInstance', pinnedContractObj.address, pinnedContractObj.abi, pinnedContractObj.name, pinnedContractObj.pinnedAt, pinnedContractObj.filePath)
```

Comment:
it would be nice if we could move this away from a react hook, basically you're looking at the API state props.plugin.REACT_API.selectExEnv, props.plugin.REACT_API.networkName in a hook but not doing anything to related to the UI except the ref enableSave which could be a state var and not a ref. it feels like this logic of saving and migrating data does not belong in a hook here. enableSave could be in a reducer if there is one or something...

Reply:
will do the refactoring in another PR